parse/ast.go

```diff
@@ -208,41 +209,32 @@ func (tp TextPattern) String() string {
 }
 func (tp TextPattern) validate() error { return nil }
 
-type PlaceholderPattern struct {
-	Pattern
-
-	Expression Expression // LiteralExpression, VariableExpression, or AnnotationExpression
-}
-
-func (pp PlaceholderPattern) String() string { return fmt.Sprint(pp.Expression) }
-
-func (pp PlaceholderPattern) validate() error {
-	if pp.Expression == nil {
-		return errors.New("placeholderPattern: expression is required")
-	}
-
-	if err := pp.Expression.validate(); err != nil {
-		return fmt.Errorf("placeholderPattern.%w", err)
-	}
-
-	return nil
-}
-
 // --------------------------------Expression----------------------------------
 
+// TODO: Reduce complexity: One expression type instead of three
+
 type LiteralExpression struct {
 	Expression
 
-	Literal    Literal    // QuotedLiteral or UnquotedLiteral
-	Annotation Annotation // Optional: FunctionAnnotation, PrivateUseAnnotation, or ReservedAnnotation
+	Literal    Literal     // QuotedLiteral or UnquotedLiteral
+	Annotation Annotation  // Optional: FunctionAnnotation, PrivateUseAnnotation, or ReservedAnnotation
+	Attributes []Attribute // Optional
 }
 
 func (le LiteralExpression) String() string {
-	if le.Annotation == nil {
+	hasAnnotation := le.Annotation != nil
+	hasAttributes := len(le.Attributes) != 0
+
+	switch {
+	case !hasAnnotation && !hasAttributes: // Only literal
 		return fmt.Sprintf("{ %s }", le.Literal)
+	case hasAnnotation && !hasAttributes: // Literal + annotation
+		return fmt.Sprintf("{ %s %s }", le.Literal, le.Annotation)
+	case !hasAnnotation && hasAttributes: // Literal + attributes
+		return fmt.Sprintf("{ %s %s }", le.Literal, sliceToString(le.Attributes, " "))
+	default: // Literal + annotation + attributes
+		return fmt.Sprintf("{ %s %s %s }", le.Literal, le.Annotation, sliceToString(le.Attributes, " "))
```

Comment:
Out of the scope of your PR. Curious.

Does empty string literal print `{  ` or `{ ||`?

Reply:
latter